Libc/net/getnetent.3
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)getnetent.3	8.1 (Berkeley) 6/4/93
.\" $FreeBSD: src/lib/libc/net/getnetent.3,v 1.16 2001/10/01 16:08:55 ru Exp $
.\"
.Dd June 4, 1993
.Dt GETNETENT 3
.Os
.Sh NAME
.Nm getnetent ,
.Nm getnetbyaddr ,
.Nm getnetbyname ,
.Nm setnetent ,
.Nm endnetent
.Nd get network entry
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In netdb.h
.Ft struct netent *
.Fn getnetent void
.Ft struct netent *
.Fn getnetbyname "const char *name"
.Ft struct netent *
.Fn getnetbyaddr "unsigned long net" "int type"
.Ft void
.Fn setnetent "int stayopen"
.Ft void
.Fn endnetent void
.Sh DESCRIPTION
The
.Fn getnetent ,
.Fn getnetbyname ,
and
.Fn getnetbyaddr
functions
each return a pointer to an object with the
following structure describing an internet network.
This structure contains either the information obtained
from the nameserver,
.Xr named 8 ,
broken-out fields of a line in the network data base
.Pa /etc/networks ,
or entries supplied by the
.Xr yp 4
system.  The order of the lookups is controlled by the
`networks' entry in
.Xr nsswitch.conf 5 .
.Pp
.Bd -literal -offset indent
struct	netent {
	char		*n_name;	/* official name of net */
	char		**n_aliases;	/* alias list */
	int		n_addrtype;	/* net number type */
	unsigned long	n_net;		/* net number */
};
.Ed
.Pp
The members of this structure are:
.Bl -tag -width n_addrtype
.It Fa n_name
The official name of the network.
.It Fa n_aliases
A zero terminated list of alternate names for the network.
.It Fa n_addrtype
The type of the network number returned; currently only AF_INET.
.It Fa n_net
The network number.  Network numbers are returned in machine byte
order.
.El
.Pp
The
.Fn getnetent
function
reads the next line of the file, opening the file if necessary.
.Pp
The
.Fn setnetent
function
opens and rewinds the file.  If the
.Fa stayopen
flag is non-zero,
the net data base will not be closed after each call to
.Fn getnetbyname
or
.Fn getnetbyaddr .
.Pp
The
.Fn endnetent
function
closes the file.
.Pp
The
.Fn getnetbyname
function
and
.Fn getnetbyaddr
sequentially search from the beginning
of the file until a matching
net name or
net address and type is found,
or until
.Dv EOF
is encountered.
The
.Fa type
must be
.Dv AF_INET .
Network numbers are supplied in host order.
.Sh FILES
.Bl -tag -width /etc/nsswitch.conf -compact
.It Pa /etc/networks
.It Pa /etc/nsswitch.conf
.It Pa /etc/resolv.conf
.El
.Sh DIAGNOSTICS
Null pointer
(0) returned on
.Dv EOF
or error.
.Sh SEE ALSO
.Xr networks 5
.Pp
.%T RFC 1101
.Sh HISTORY
The
.Fn getnetent ,
.Fn getnetbyaddr ,
.Fn getnetbyname ,
.Fn setnetent ,
and
.Fn endnetent
functions appeared in
.Bx 4.2 .
.Sh BUGS
The data space used by
these functions is static; if future use requires the data, it should be
copied before any subsequent calls to these functions overwrite it.
Only Internet network
numbers are currently understood.
Expecting network numbers to fit
in no more than 32 bits is probably
naive.


X11/ChangeLog
2003-09-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: revert fix for
	3363492 and add another check for RootlessKnowsWindowNumber ()
	to fix the original bug (3420745)

2003-09-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c (sigchld_handler):
	loop until no more zombies exist (3414442)

2003-08-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m,
	xc/programs/Xserver/hw/apple/X11Application.h: move calling
	HideMenuBar and ShowMenuBar onto the main UI thread (3359753)

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	few more MT changes, be even more careful about when contexts
	are updated

2003-08-07  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	Large set of changes to make this code handle multiple threads
	sanely. Changes include adding a mutex for each screen, and
	only ever touching CGL contexts when we know they're bound to
	the current thread or no thread. This involves patching gli
	dispatch table to cause glViewport to call back into DRI code
	so that contexts can be updated on their own thread. May be
	occasions where glXSwapBuffers fails if no suitable context can
	be found to issue the swap on (3340377) 

	* xc/programs/Xserver/hw/apple/quartz.c: quartzServerVisible
	variable needs to start out as false (3366845)

2003-08-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: move the call
	to RootlessKnowsWindowNumber () into the event handler so that
	we don't activate X on the broken menu-mouse-down event
	(3363492)

	* RELEASE-NOTES: updated to reflect that X11 is now part of the
	OS and no longer a preview release (3361286)

2003-07-29  John Harper  <jsh@apple.com>

	* xc/lib/GL/glx/glxext.c (glXMakeCurrent): unlock display after
	sending GLXMakeCurrent request when changing from indirect to
	direct context (3351489, sent upstream via Torrey)

2003-07-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	when in fullscreen mode, don't pass ApplicationDidReactivate
	events through to AppKit (3341360)

2003-07-23  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.c (XAppleDRIGetIndirectContext):
	make the context we allocate 1k bigger than we think it needs
	to be, in case GL changes

2003-07-22  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.c (XAppleDRIGetIndirectContext):
	allocate the CGLContext used for dispatch indirect rendering
	ourselves, avoids ever needing to contact the window server
	(3126715)

2003-07-18  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/dri-surface.c: fixed use of
	pixmap private index

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzSetFullscreen):
	use xp_disable_hot_keys instead of CG spi (3221162)

	* xc/programs/Xserver/hw/apple/rootless-screen.c,
	xc/programs/Xserver/hw/apple/rootless-common.h,
	xc/programs/Xserver/hw/apple/rootless-common.c: use
	RootlessFinishedDrawing () to call new function
	RootlessMayRedisplay (). After a certain amount of time since
	the last update it will cause a flush (3270678)

	* xc/programs/Xserver/hw/apple/rootless-screen.c: clean-up

2003-07-17  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessPaintWindowBackground): prevent trying to draw in
	surfaces here

	* xc/programs/Xserver/hw/apple/rootless-gc.c: call
	RootlessFinishedDrawing at the end of each rendering primitive

	* xc/programs/Xserver/hw/apple/rootless-common.h,
	xc/programs/Xserver/hw/apple/rootless-common.c: use the
	DRIStartDrawing functions etc to test if the drawable is a
	surface.
	(RootlessFinishedDrawing): new function, should be called after
	drawing has finished

	* xc/programs/Xserver/hw/apple/dristruct.h,
	xc/programs/Xserver/hw/apple/dri-surface.h,
	xc/programs/Xserver/hw/apple/dri-surface.c (DRIStartDrawing,
	DRIStopDrawing, DRIDamageRegion, DRISynchronizeDrawable,
	DRISynchronize): functions to support Xlib rendering on
	accelerated surfaces. Only supported on 10.3, and they won't do
	anything by default

	* xc/programs/Xserver/hw/apple/dri-surface.c
	(DRICreateSurface): support Pixmap drawables (3122756, 10.3
	only)

	* xc/programs/Xserver/hw/apple/dri-surface.c: removed check for
	cg from SUJagGreen

	* xc/programs/Xserver/GL/apple/indirect.c: support offscreen
	drawables (by removing the code that stopped them being
	supported, dri-surface.c does the real work) (3122756)

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/quartz-server.c,
	xc/programs/Xserver/hw/apple/X11Application.m,
	xc/programs/Xserver/hw/apple/X11Application.h: added xp_options
	default, lets you pass extra init options to Xplugin

	* xc/programs/Xserver/hw/apple/*: fixed gcc 3.3 warnings

2003-07-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/x-list.c,
	xc/programs/Xserver/hw/apple/x-list.h (x_list_pop): new
	function

	* xc/programs/Xserver/hw/apple/x-hash.c: fix gcc 3.3 warning

2003-07-07  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: don't pass
	flags changed events to X11 when Appkit is active (3316333)

2003-07-02  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzRunClient):
	don't quote commands, it stops multi-word commands working

	* xc/programs/Xserver/hw/apple/X11Controller.m: updated quit
	alert text (3313133)

	* xc/programs/Xserver/hw/apple/Xquartz.man: added more defaults

	* xc/programs/Xserver/hw/apple/bundle-main.c (main): detach
	from the controlling terminal when we have one (avoids bash
	suspending the process when we try to launch problems!?)

2003-06-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/applewm-impl.c
	(ProcAppleWMSetWindowLevel): removed the code to call

Libinfo/dns.subproj/getnetent.c
/*
 * Copyright (c) 1999 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * Portions Copyright (c) 1999 Apple Computer, Inc.  All Rights
 * Reserved.  This file contains Original Code and/or Modifications of
 * Original Code as defined in and that are subject to the Apple Public
 * Source License Version 1.1 (the "License").  You may not use this file
 * except in compliance with the License.  Please obtain a copy of the
 * License at http://www.apple.com/publicsource and read it before using
 * this file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE OR NON- INFRINGEMENT.  Please see the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/* Portions Copyright (c) 1993 Carlos Leandro and Rui Salgueiro
 *	Dep. Matematica Universidade de Coimbra, Portugal, Europe
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * from getnetent.c	1.1 (Coimbra) 93/06/02
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)getnetent.c	8.1 (Berkeley) 6/4/93";
static char rcsid[] = "$Id: getnetent.c,v 1.4 2003/02/18 17:29:23 majka Exp $";
#endif /* LIBC_SCCS and not lint */

#include <sys/param.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <arpa/nameser8_compat.h>

#include <stdio.h>
#include <stdlib.h>
#include <resolv8_compat.h>
#include <netdb.h>
#include <string.h>

#ifndef _PATH_NETWORKS 
#define _PATH_NETWORKS  "/etc/networks"
#endif

#define	MAXALIASES	35

static FILE *netf;
static struct netent net;
static char *net_aliases[MAXALIASES];
#if defined(__APPLE__)
extern
#endif
int _net_stayopen;

void _setnetent __P((int));
void _endnetent __P((void));

void
setnetent(stayopen)
	int stayopen;
{

	sethostent(stayopen);
	_setnetent(stayopen);
}

void
endnetent()
{

	endhostent();
	_endnetent();
}

void
_setnetent(f)
	int f;
{

	if (netf == NULL)
		netf = fopen(_PATH_NETWORKS, "r" );
	else
		rewind(netf);
	_net_stayopen |= f;
}

void
_endnetent()
{

	if (netf) {
		fclose(netf);
		netf = NULL;
	}
	_net_stayopen = 0;
}

struct netent *
getnetent()
{
	char *p;
	static char *line = NULL;
	register char *cp, **q;

	if (netf == NULL && (netf = fopen(_PATH_NETWORKS, "r" )) == NULL)
		return (NULL);

	if (line == NULL) {
		line = malloc(BUFSIZ+1);
		if (line == NULL)
			return (NULL);
	}
again:
	p = fgets(line, BUFSIZ, netf);
	if (p == NULL)
		return (NULL);
	if (*p == '#')
		goto again;
	cp = strpbrk(p, "#\n");
	if (cp == NULL)
		goto again;
	*cp = '\0';
	net.n_name = p;
	cp = strpbrk(p, " \t");
	if (cp == NULL)
		goto again;
	*cp++ = '\0';
	while (*cp == ' ' || *cp == '\t')
		cp++;
	p = strpbrk(cp, " \t");
	if (p != NULL)
		*p++ = '\0';
	net.n_net = inet_network(cp);
	net.n_addrtype = AF_INET;
	q = net.n_aliases = net_aliases;
	if (p != NULL) 
		cp = p;
	while (cp && *cp) {
		if (*cp == ' ' || *cp == '\t') {
			cp++;
			continue;
		}
		if (q < &net_aliases[MAXALIASES - 1])
			*q++ = cp;
		cp = strpbrk(cp, " \t");
		if (cp != NULL)
			*cp++ = '\0';
	}
	*q = NULL;
	return (&net);
}
